### PR TITLE
color sessions: add function to compare uidisplaydata agent tree structures

### LINKED ISSUE #581: function to compare if agent tree structures match
## Use Case

Advances #511 

To manage applying stored color settings, we need to be able to confirm that two trajectories (which may have the same name) in fact have identical matching agent trees, with possibly different color assignments. Otherwise, we may attempt to apply incompatible color settings.

## Acceptance Criteria

A util function with unit tests that can take in instances of `UIDisplayData` and verify that the agents and displayStates are identical between the two.

## Details

_Please provide any helpful specifications_

### PR DESCRIPTION
**Pull request by interim17**
Time estimate or Size
=======
_small, mostly tests_

Problem
=======
Closes #581, advances #511

When retrieving stored color settings from browser storage, we will retrieve data according to trajectory name. It's possible for incompatible trajectory data structures to have the same name, so we need a utility to compare the agents and displayStates to ensure that the retrieved settings can be safely applied to the trajectory that is being loaded.

This PR is just the utility and unit tests, not an implementation of retrieving settings.

### CODE REVIEW

**Review by meganrm — APPROVED** [review]

**Review by tyler-foster — APPROVED** [review]
LGTM.

Is it smelly for this to exist? Seems a little curious, but I'd need a better understanding of how `UIDisplayData` gets used to really comment on it.

File: src/util/index.ts [hunk 1]
```diff
@@ -105,3 +106,26 @@ export const roundToTimeStepPrecision = (
     const multiplier = Math.pow(10, precision);
     return Math.round(input * multiplier) / multiplier;
 };
+
+export const compareAgentTrees = (a: UIDisplayData, b: UIDisplayData) => {
```
[the remark this answers is not in the document]
Comment (interim17) [reply src/util/index.ts:110]: Sure I can add one.

**Comment by interim17** [comment]
> LGTM.
> 
> Is it smelly for this to exist? Seems a little curious, but I'd need a better understanding of how `UIDisplayData` gets used to really comment on it.

I'm not very familiar with this use of the word smelly.

Is it smelly because it doesn't get used yet, so the purpose isn't clear? Or smelly because you don't know what if `UIDisplayData` is a good data structure for this use case?

For the latter question, I'll just say that `UIDisplayData` is a tree of agents and displayStates with associated color assignments. The viewer generates a tree of that type when parsing a trajectory file and passes it to the website to use, the website can either mutate it or generate new trees when changing color assignments.

@meganrm and I paired on this issue a while back and landed on `UIDisplayData` as a good existing data structure to store a snapshot of the current state of color assignments for a trajectory. Happy to give more context, and definitely open to input/feedback if you see potential issues here.

**Comment by tyler-foster** [comment]
> I'm not very familiar with this use of the word smelly.

As in, is this a "code smell"?

> Is it smelly because it doesn't get used yet, so the purpose isn't clear? Or smelly because you don't know what if `UIDisplayData` is a good data structure for this use case?
> 
> For the latter question, I'll just say that `UIDisplayData` is a tree of agents and displayStates with associated color assignments. The viewer generates a tree of that type when parsing a trajectory file and passes it to the website to use, the website can either mutate it or generate new trees when changing color assignments.
> 
> @meganrm and I paired on this issue a while back and landed on `UIDisplayData` as a good existing data structure to store a snapshot of the current state of color assignments for a trajectory. Happy to give more context, and definitely open to input/feedback if you see potential issues here.

Yessir, I meant the latter question. My concern stems from my lack of understanding of this part of your PR description:

> It's possible for incompatible trajectory data structures to have the same name, so we need a utility

It sounds like we've arrived at a namespacing issue that now begets a special solution, rather than a fix to the namespacing issue. I'd have to have a better understanding of what it means in practice for the trajectory data structures to have the same name before I could provide an alternative idea - you guys could be totally correct in how you've solved this problem though. I say just keep moving forward with this solution and if it ends up feeling awkward for some reason you can always double back and change it again.

**Comment by interim17** [comment]
> It sounds like we've arrived at a namespacing issue that now begets a special solution, rather than a fix to the namespacing issue. I'd have to have a better understanding of what it means in practice for the trajectory data structures to have the same name before I could provide an alternative idea. You guys could be totally correct in how you've solved this problem though. I say just keep moving forward with this solution and if it ends up feeling awkward for some reason you can always double back and change it again.

Ok I see where you're at.

When we store color settings in browser storage the key we use is the filename. Say its "test". If we have {test: settings} stashed, and load a different trajectory, but also named "test", trying to apply the stashed settings will cause an error. This is an edge case, because names are likely to be unique, and this storage is ephemeral.

I agree I think its ok to go forward for now, so I'm going to merge, but if you have suggestions on a better way to handle the namespace in the browser storage, I'm open to it. Ultimately this is a convenience for temporarily storing/retrieving ongoing adjustments to colors and the more permanent feature, down the line, will be writing changes to a new trajectory file.
